### ISSUE TITLE
Add option to extract simple project meta-data (name & version)

### ISSUE THREAD

**Issue by exhuma** [pypa/build#348]
When using setuptools with a `setup.py` I can run `python setup.py --version` and `setup.py --fullname` which are extremely useful in CI pipelines (and other automated tasks). They can be run without actually *building* the distribution first.

PEP-517 abstracts away the build-system and there is no deterministic way anymore to extract the project's version number during the build-process (at least not as far as I can tell).

As different build-systems store the version numbers in different locations, a well-known way to retrieve the version string would be beneficial.

This seems fairly simple to provide by exposing `pep517.meta.load().version`

Would this fit into `pyp/build`?

**Comment by FFY00** [comment]
See #340.

**Comment by FFY00** [comment]
0.7.0 has been released with support for this. Let me know if it fits your needs.

https://pypa-build--353.org.readthedocs.build/en/353/api.html#build-util-module

**Comment by exhuma** [comment]
What's the difference of this and `pep517.meta.load`? In terms of programmatic API `pep517.meta.load` gives me already what I need. But it would be more convenient to expose that via the CLI.  I realised just now that my original request was not clear enough on this aspect.

We use build and deploy automation for our projects. And right now I have developed a small in-house helper script the basically just wraps `pep517.meta.load` in a CLI executable.

I think that the use-case to determine the project meta-data (especially name and version) are extremely common in CI pipelines (or other build/deploy instrumentation). For that reason, having the possibility to just run `pyproject-build --pkg-version` or `pyproject-build --pkg-name` would be super helpful.

Basically reproducing the same behaviour as `python setup.py --version` and `python setup.py --name`.

Back in the days when distutils and setuptools were basically the main way to go, these two commands served me very well.

But now, with an ever more diverse build toolchain (thanks to PEP-517) this use-case of "getting the project-version and name" has become more opaque. And adding this to a CLI executable like `pyproject-build` would be beneficial.

Aside from project-name and -version all the other meta-data are of course also very valuable (like dependencies). `pep517.meta.load` already returns them. They are just not exposed by `pyproject-build`.

**Comment by FFY00** [comment]
Okay, I understand now, sorry. I don't think it is in scope for this project, as we should limit our CLI/API surface due to this package being critical for bootstrapping Python environments, and that functionality is non-trivial.

We can, however, create a helper package for that. Something like `build-metadata`, which would expose a `pyproject-metadata` command.

**Comment by layday** [comment]
> I think that the use-case to determine the project meta-data (especially name and version) are extremely common in CI pipelines (or other build/deploy instrumentation). For that reason, having the possibility to just run `pyproject-build --pkg-version` or `pyproject-build --pkg-name` would be super helpful.

CI pipelines generally offer some way to manipulate JSON, e.g. with a function in the style of GitHub's `fromJSON` or by bootstrapping `jq`.  I think the way to go here would be to create a library to transform the distribution metadata into the JSON format specified in [PEP 566](https://www.python.org/dev/peps/pep-0566/) instead of designing yet another interface for working with these metadata tailored to the CLI.

**Comment by exhuma** [comment]
I think having *some* way to extract the most basic meta-data would be beneficial. There are many CI tasks which could benefit from that. For example (and this is the origin of my request), we have a "build-docs" step which needs project name and version for file/folder-names. And those always include the name and version.

By switching over to full PEP-517 with `pyproject-build` I will lose that information. Or more precisely, I will no longer have an easy way to access that info. I would need to create a small script <10 lines just to import pep517.meta.load and fetch the data. And that script would need to be available in each of the pipelines.

**Comment by FFY00** [comment]
I mean, you could do something like the following

```
python -c 'import pep517.meta; m = pep517.meta.load(); print(f"{m.name}=={m.version}")'
```

And nowadays, several CI systems allow using Python as shell

https://github.com/pypa/build/blob/db905dbee2225e415fd252bae7ede7ef3b70fe2d/.github/workflows/test.yml#L46-L60

But like I mentioned above, if there is indeed a need for this, I would be happy to provide it as an external package.

**Comment by exhuma** [comment]
I understand if this is deemed to be "out-of-scope" of the project. And there *is* a workaround so for me it's okay if this is either closed or put "on-hold" to see if there's interest by more than just me 😄

**Comment by FFY00** [comment]
My concern about the external package is understanding exactly the use-cases people have, so if anyone has a similar use-case, please let me know.

**Comment by FFY00** [comment]
FYI I have started working on something, it will also read the [PEP 621](https://www.python.org/dev/peps/pep-0621/#name) metadata instead if available.

https://github.com/FFY00/python-metadata

I am closing this issue here as out of scope.

**Comment by sinoroc** [comment]
What is the current advice for such a thing?

I am not sure what to tell people who ask for this:
* https://stackoverflow.com/q/75373164
* https://stackoverflow.com/q/71271825
* https://stackoverflow.com/q/71592060
* https://stackoverflow.com/q/73506836

**Comment by FFY00** [comment]
There isn't any tool that implements that AFAIK, but it shouldn't be very hard to make one. I haven't had time to keep up with https://github.com/FFY00/python-metadata.

I think the best option at the moment is
```
python -c "import build.util; print(build.util.project_wheel_metadata('.')['Version'])"
```

**Comment by FFY00** [comment]
I just noticed that you recommend that too :sweat_smile:

**Comment by sinoroc** [comment]
OK, thanks :)

**Comment by FFY00** [comment]
If you want to work on https://github.com/FFY00/python-metadata let me know btw. I may have time to discuss and review things, if there are people interested, but not to lead the development.

**Comment by exhuma** [comment]
@sinoroc since I opened this ticket I tried various variants. The only one I found acceptable was using `tomlq` (part of [yq](https://github.com/mikefarah/yq)), giving me the following command to extract the version:

```
tomlq -r .project.version pyproject.toml
```

This is [PEP-621](https://peps.python.org/pep-0621/#version) compliant. That PEP dictates that the `version` key should be in the `project` table. So `.project.version` gets at that key. This will not work if you use a tool that's not PEP-621 compliant (like poetry and flit for example).

The advantage of `yq` is that it's *fast*. It only needs to open the toml file to read the value.

Any tool which leverages [PEP-571](https://peps.python.org/pep-0571) will inevitably be slow. This is because it will need to create a new build-environment and install the build-system. Only then can it use the build system to get the version. The advantage of *that* is that it is totally independent and will work with tools which are not fully PEP-621 compliant.

Personally, I am now exclusively using `tomlq`. I use `setuptools` for all my projects so I can safely assume that `.project.version` is available. If I encounter a project that's different, then calling `tomlq` with a different query is not all that much of a hassle.

And at least with that I have something that's *fast*, which is important when you just want to run a quick shell-command like:

```
git tag -m Release v$(tomlq -r .project.version pyproject.toml)
```

**Comment by sinoroc** [comment]
For completeness:

With Python 3.11+ you can use `tomllib` and write a one-liner
... if `version` is not `dynamic`.

**Comment by sinoroc** [comment]
> Is it secretly building an entire .whl file somewhere behind the scenes?

As far as I recall, yes. Or something close to a wheel. If I am not mistaken, in theory there is an API specific for "I want metadata only", but not many build backends implement it so they kind of build the whole thing instead. Because yes the command you show is build backend-agnostic, i.e. it will work with any build backends. Whereas `python setup.py --version` is obviously specific to setuptools, and some other tools have similar proprietary solution, I think `poetry version` for example (which I assume also answers very fast).

**Comment by layday** [comment]
The API you allude to is the `prepare_metadata_for_build_wheel` hook, which is indeed optional, and which all major backends have implemented.  A wheel is rarely built, but you're still bearing the cost of:

* Installing the backend's dependencies
* Importing the backend in a subprocess
* Calling the hook
* Reading the metadata from the filesystem and parsing it as an email header

**Comment by henryiii** [comment]
It's also really hard to implement reliably. Backends often disable this for special situations. scikit-build-core disables it if you have an override on build failure - the metadata then in that case could depend on if your code fails to compile. Hatching disabled it for pip: https://github.com/pypa/hatch/blob/5352e4422636cf1238017a74f0c67d689ccee558/backend/src/hatchling/build.py#L86-L98

If you have static metadata and use PEP 621, pyproject-metdata can read this, maybe we should expose the json output via a CLI. It won't work for dynamic metadata, though. That could be a fast-path.

**Comment by sinoroc** [comment]
> If you have static metadata and use PEP 621, pyproject-metdata can read this

Note also that, for simple cases like reading the static version, and with Python 3.11+, one can probably use [tomllib](https://docs.python.org/3/library/tomllib.html). No need for external dependencies. Can probably be made into a one-liner. I have not tried, so I hope I am not telling lies.

**Comment by henryiii** [comment]
Yes, that would be
```
python3 -c 'import tomllib, pathlib; print(tomllib.loads(pathlib.Path("pyproject.toml").read_text(encoding="utf-8"))["project"]["version"])'
```
